common/src/ethereum/erc_20.rs

```diff
@@ -0,0 +1,148 @@
+}
+
+impl<M> ERC20<M>
+where
+    M: Middleware,
+    Error<M>: From<<M as Middleware>::Error>,
+{
+    pub fn new(client: impl Into<Arc<M>>, address: Address) -> Self {
+        let abi: Abi = serde_json::from_str(ABI).expect("The ABI is valid");
+        let contract = Contract::new(address, abi, client);
+
+        Self { contract }
+    }
+
+    pub async fn get_symbol(&self) -> Result<String, Error<M>> {
+        let symbol = self
+            .contract
+            .method("symbol", ())
+            .map_err(ContractError::from)?
+            .call()
+            .await?;
+
+        Ok(symbol)
+    }
+
+    pub async fn get_decimals(&self) -> Result<u8, Error<M>> {
+        let symbol = self
+            .contract
+            .method("decimals", ())
+            .map_err(ContractError::from)?
+            .call()
+            .await?;
+
+        Ok(symbol)
+    }
+
+    pub async fn get_allowance(&self, owner: Address, spender: Address) -> Result<U256, Error<M>> {
+        let allowance = self
+            .contract
+            .method("allowance", (owner, spender))
+            .map_err(ContractError::from)?
+            .call()
+            .await?;
+
+        Ok(allowance)
+    }
+
+    pub fn approve(&self, spender: Address, amount: U256) -> Result<ContractCall<M, ()>, Error<M>> {
+        self.contract
+            .method("approve", (spender, amount))
+            .map_err(Error::ContractAbi)
+    }
+}
+
+#[derive(Debug)]
+pub enum Token {
+    WETH,
+    DAI,
+    USDC,
+    USDT,
+    Other(Address), // e.g 0x2b591e99afe9f32eaa6214f7b7629768c40eeb39
+}
+
+impl FromStr for Token {
+    type Err = anyhow::Error;
+
+    fn from_str(token: &str) -> Result<Token, Self::Err> {
+        match token.to_uppercase().as_str() {
+            "WETH" => Ok(Token::WETH),
+            "DAI" => Ok(Token::DAI),
+            "USDC" => Ok(Token::USDC),
+            "USDT" => Ok(Token::USDT),
+            _ => {
+                if let Ok(address) = Address::from_str(token) {
+                    Ok(Token::Other(address))
+                } else {
+                    Err(anyhow::anyhow!(
+                        "Failed to parse token's address: {}",
+                        token
+                    ))
+                }
+            }
+        }
+    }
+}
+
+impl Token {
+    pub fn get_decimals(&self) -> Option<usize> {
+        match self {
+            Token::WETH => Some(18),
+            Token::DAI => Some(18),
+            Token::USDC => Some(6),
+            Token::USDT => Some(6),
+            Token::Other(_) => None,
+        }
+    }
+
+    pub fn get_address(&self) -> Option<Address> {
+        match self {
+            Token::WETH => Some(*WETH_ADDRESS),
+            Token::DAI => Some(*DAI_ADDRESS),
+            Token::USDC => Some(*USDC_ADDRESS),
+            Token::USDT => Some(*USDT_ADDRESS),
+            Token::Other(address) => Some(*address),
+        }
```

Comment:
This can return `Address` now

Reply:
https://github.com/radicle-dev/radicle-cli/compare/f4007e459f8ce654f3ca58ce48dc1f9910838534..a664622f7600c99dc0622b776816712d0b049d45#diff-5b91c481d0947d833cc80f7e2abb8a4f92ee01e5f225e0c526772affd43acc47R139